File: CryptoLib/CryptoLib/XmlParser/CdocParserDelegate.m
//
//  CdocParserDelegate.m
//  CryptoLib
/*
 * Copyright 2017 - 2021 Riigi Infosüsteemi Amet
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#import "CdocParserDelegate.h"
@interface CdocParserDelegate ()  <NSXMLParserDelegate>
@end

@implementation CdocParserDelegate {
}

- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName attributes:(NSDictionary *)attributeDict {
    if ([elementName isEqualToString:@"denc:EncryptedKey"]) {
        if (_addressees == nil){
            _addressees  = [NSMutableArray new];
        }
        if (_lastAddressee == nil){
            _lastAddressee  = [Addressee new];
        }
        NSString *attribute = attributeDict[@"Recipient"];
        NSArray *cn = [attribute componentsSeparatedByString:@","];
        Addressee *addressee = [Addressee new];
        if (cn.count > 1) {
            addressee.surname = cn[0];
            addressee.givenName = cn[1];
            addressee.identifier = cn[2];
        } else {
            addressee.identifier = cn[0];
        }
        [_addressees addObject:addressee];
        _lastAddressee = addressee;
    }
    if ([elementName isEqualToString:@"ds:X509Certificate"]) {
        _isNextCharactersCertificate = YES;
    }
    if ([elementName isEqualToString:@"denc:EncryptionProperty"] && [[attributeDict valueForKey: @"Name"]  isEqual: @"orig_file"]) {
        _isNextCharactersFilename = YES;
    }
}

-(void) parser:(NSXMLParser *)parser foundCharacters:(NSString *)string {
    if (_isNextCharactersFilename) {
        if (_currentFilenameNode) {
            _currentFilenameNode = [_currentFilenameNode stringByAppendingString:string];
        } else {
            _currentFilenameNode = string;
        }
    }
        
    if (_isNextCharactersCertificate) {
        NSString *trimmedString = [string stringByTrimmingCharactersInSet:[NSCharacterSet newlineCharacterSet]];
        NSString *pemCertificate = [NSString stringWithFormat:@"-----BEGIN CERTIFICATE-----\n%@\n-----END CERTIFICATE------", trimmedString];
        
        NSData *pemFormattedCertificate = [pemCertificate dataUsingEncoding:NSUTF8StringEncoding];
        _lastAddressee.cert = [pemFormattedCertificate subdataWithRange:NSMakeRange(0, [pemFormattedCertificate length] - 1)];
    }
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(nullable NSString *)namespaceURI qualifiedName:(nullable NSString *)qName {
    if (_isNextCharactersFilename) {
        if (_dataFiles == nil){
            _dataFiles  = [NSMutableArray new];
        }
        NSArray *filenameWithBytesLength = [_currentFilenameNode componentsSeparatedByString:@"|"];
        NSString *filename = filenameWithBytesLength[0];
        CryptoDataFile *dataFile = [CryptoDataFile new];
        dataFile.filename = filename;
        [_dataFiles addObject:dataFile];
    }
    _currentFilenameNode = nil;
    _isNextCharactersFilename = NO;
    _isNextCharactersCertificate = NO;
}

@end

File: CryptoLib/CryptoLib/XmlParser/CdocParserDelegate.swift
//
//  CdocParserDelegate.swift
//  CryptoLib
/*
 * Copyright 2017 - 2024 Riigi Infosüsteemi Amet
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

import Foundation

public class CdocParser: NSObject {
    @objc static public func parse(path: String) -> CdocInfo? {
        guard let parser = XMLParser(contentsOf: URL(fileURLWithPath: path)) else {
            NSLog("Error: Unable to read file at \(path)")
            return nil
        }
        let delegate = CdocParserDelegate()
        parser.delegate = delegate;
        guard parser.parse() else {
            NSLog("Error: Failed to parse XML")
            return nil
        }
        return CdocInfo(addressees: delegate.addressees, dataFiles: delegate.dataFiles)
    }
}

class CdocParserDelegate: NSObject, XMLParserDelegate {
    public var addressees: [Addressee] = []
    public var dataFiles: [CryptoDataFile] = []
    var data: String? = nil

    func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String: String]) {
        switch elementName {
        case "ds:X509Certificate":
            data = String()
        case "denc:EncryptionProperty" where attributeDict["Name"] == "orig_file":
            data = String()
        default: break
        }
    }

    func parser(_ parser: XMLParser, foundCharacters string: String) {
        if data != nil {
            data! += string
        }
    }

    func parser(_ parser: XMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        guard data != nil else { return }
        switch elementName {
        case "ds:X509Certificate":
            if let data = Data(base64Encoded: data!, options: .ignoreUnknownCharacters) {
                addressees.append(Addressee(cert: data))
            }
        case "denc:EncryptionProperty":
            if let filename = data!.split(separator: "|").first {
                dataFiles.append(CryptoDataFile(filename: String(filename)))
            }
        default: break
        }
        data = nil
    }
}
